lang/src/accounts/account_loader.rs

```diff
@@ -116,51 +114,61 @@ impl<'info, T: ZeroCopy + Owner> AccountLoader<'info, T> {
+        // to init an account from a different program through a CPI and then
+        // read its data once the CPI returns. Therefore, the caller of the CPI
+        // shouldn't try to initialize the account
+        // because it's not owned by them.
+        if T::owner() == *program_id {
+            let data: &mut [u8] = &mut info.try_borrow_mut_data()?;
+            if data.len() < 8 {
+                return Err(anchor_lang::error::ErrorCode::AccountDidNotSerialize.into());
+            }
+            crate::solana_program::program_memory::sol_memcpy(data, &T::DISCRIMINATOR, 8);
         }
-        let data: &[u8] = &acc_info.try_borrow_data()?;
-        if data.len() < T::discriminator().len() {
+        Self::try_from_unchecked(program_id, info)
+    }
+
+    /// Constructs a new `AccountLoader` from a previously initialized account.
+    #[inline(never)]
+    pub fn try_from(info: &AccountInfo<'info>) -> Result<AccountLoader<'info, T>> {
+        Self::account_checks(info)?;
+
+        let data: &[u8] = &info.try_borrow_data()?;
+        if data.len() < T::DISCRIMINATOR.len() {
             return Err(ErrorCode::AccountDiscriminatorNotFound.into());
         }
         // Discriminator must match.
         let disc_bytes = array_ref![data, 0, 8];
-        if disc_bytes != &T::discriminator() {
+        if disc_bytes != &T::DISCRIMINATOR {
             return Err(ErrorCode::AccountDiscriminatorMismatch.into());
         }
 
-        Ok(AccountLoader::new(acc_info.clone()))
+        Ok(AccountLoader::new(info.clone()))
     }
 
-    /// Constructs a new `Loader` from an uninitialized account.
+    /// Constructs a new `AccountLoader` without checking the discriminator.
     #[inline(never)]
     pub fn try_from_unchecked(
         _program_id: &Pubkey,
-        acc_info: &AccountInfo<'info>,
+        info: &AccountInfo<'info>,
     ) -> Result<AccountLoader<'info, T>> {
-        if acc_info.owner != &T::owner() {
+        Self::account_checks(info)?;
+        Ok(AccountLoader::new(info.clone()))
+    }
+
+    fn account_checks(info: &AccountInfo<'info>) -> Result<()> {
+        if info.owner != &T::owner() {
             return Err(Error::from(ErrorCode::AccountOwnedByWrongProgram)
-                .with_pubkeys((*acc_info.owner, T::owner())));
+                .with_pubkeys((*info.owner, T::owner())));
         }
-        Ok(AccountLoader::new(acc_info.clone()))
+        Ok(())
     }
 
     /// Returns a Ref to the account data structure for reading.
     pub fn load(&self) -> Result<Ref<T>> {
         let data = self.acc_info.try_borrow_data()?;
-        if data.len() < T::discriminator().len() {
-            return Err(ErrorCode::AccountDiscriminatorNotFound.into());
-        }
-
-        let disc_bytes = array_ref![data, 0, 8];
-        if disc_bytes != &T::discriminator() {
-            return Err(ErrorCode::AccountDiscriminatorMismatch.into());
-        }
-
```

Comment:
Im almost but not 100% sure we can remove these checks. `init` and `zero` both set the discriminator instead of `load_init` (with which the checks in `load` and `load_mut` mattered cause you could forget initializing with `load_init`).

`try_from` checks that the discriminator when `init` or `zero` are not used